src/CompiledMatcher.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\ConstraintInterface;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private $cacheDir;
+
+    private $checked;
+
+    public function __construct($cacheDir = null)
+    {
+        $this->cacheDir = $cacheDir ? $cacheDir : \sys_get_temp_dir();
+    }
+
+    /**
+     * Evaluates the expression: $constraint match $operator $version
+     *
+     * @param ConstraintInterface $constraint
+     * @param int                 $operator
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        $cacheKey = $operator.$constraint;
+        if (!isset($this->checked[$cacheKey])) {
+            $sha = \sha1($cacheKey);
+            $file = $this->cacheDir.'/constraint_'.$sha;
+            $function = 'match_'.$sha;
+            if (!\is_file($file)) {
+                \file_put_contents($file, '<?php function '.$function.'($v, $b){return '.$constraint->compile($operator).';}');
```

Comment:
Is there a big gain in caching to disk? Is it really worth it vs creating lambdas via eval() for example?

Reply:
we should at least keep a mode without `eval`, to account for hosts disabling it.

Reply:
Hosts without eval can just run without this optimization. Updates should not be run on these hosts anyway. The main question to me is whether writing tons of files helps or not, if it's not worth a significant gain I'd rather avoid touching the filesystem in that way.

Reply:
I'm also afraid of writing so many file on disk and understand what you mean.
Just for the record, it's 1 file per evaluated constraint (ie. `^5.0`, `^5.1` are 2 files)
If somebody have tips to avoid writing so much files.

I run composer up on magento project with 3 implementations:
- [composer original](https://blackfire.io/profiles/a74958be-1823-4364-a8a5-5921778fe09f/graph)
- [with file](https://blackfire.io/profiles/d97a1207-65dd-433b-a466-d1c64f8aa076/graph) => [-20%](https://blackfire.io/profiles/compare/0a7269b7-9e1f-48e3-b88c-8ac9ff08f048/graph)
- [with eval](https://blackfire.io/profiles/3298571c-222b-4950-9d69-146fae5b0f2f/graph) => [-1%](https://blackfire.io/profiles/compare/a504a829-47af-4061-bfdc-6ec4869d6a40/graph)

The difference is: Eval requires to recompile constraints every times.

But IMHO **those number are temporary**, this open the field to write more complex optimization on MultiConstraints.

Reply:
@Seldaek this would require making the CompiledMatcher optional, which is not the case currently. And `composer install` also runs the solver. Even if we don't need the optimization in that case, we need the matcher to work.

Reply:
@stof you just need a little more imagination.. this function could `return $constraint->match(new Constraint($op, $version))` if `eval` is not callable. Not such a big deal. Anyway this is obviously solvable one way or another, not worth discussing at this point.

@jderusse is your eval case just evaling the compiled constraint? Or evaling a function if none exists yet, and calling it if it was created already? Because I see 310730 calls to CompiledMatcher and 310730 eval calls in your profile, so it seems to me like you called eval every time, which then really bypasses any optimization potential. Ideally we'd just create an in-memory array of functions instead of dumping them to filesystem. I understand this means the compiled constraints can not be reused across runs, but I'd like to really understand if reusing them is *that* beneficial. I can also try myself when I find a minute though..

Reply:
@Seldaek My eval compiled the code once, then evaluated each time.

```php
        $cacheKey = $operator.$constraint;
        if (!isset($this->checked[$cacheKey])) {
            $code = $this->checked[$cacheKey] = '$r='.$constraint->compile($operator).';';
        } else {
            $code = $this->checked[$cacheKey];
        }

        $v = $version;
        $b = $v[0] === 'd' && 'dev-' === substr($v, 0, 4);

        eval($code);

        return $r;
```

Following your suggestion, I evaluated it once and reuse the function
```php
        $cacheKey = $operator.$constraint;
        if (!isset(static::$checked[$cacheKey])) {
            $sha = \sha1($cacheKey);
            $function = 'composer_semver_constraint_'.$sha;
            static::$checked[$cacheKey] = $function;
            eval('function '.$function.'($v, $b){return '.$constraint->compile($operator).';}');
            $function = '\\'.$function;
        } else {
            $function = static::$checked[$cacheKey];
        }

        $v = $version;

        return $function($version, $v[0] === 'd' && 'dev-' === substr($v, 0, 4));
```
and get better result. The [eval function is almost as fast as](https://blackfire.io/profiles/compare/4d9908a3-ad07-4d43-8994-1fda87b7fb8b/graph) the file one.
Letting me think to use the eval in this PR. 

When we'll improve the MultiConstraint optimize function and the ratio "time saved with compiled"/"time to compile" will decrease we'll can think about using file

Reply:
Ok that's great news thanks for testing again. You could probably even skip the sha1 call by doing `static::$checked[$cacheKey] = $function = eval('return function ($v, $b) { return ...; };')` so that would create anonymous functions.

Something like https://3v4l.org/l7YlZ in case it's not clear.

Reply:
Clever !!